[InstCombine] canonicalize sign bit checks

jacobly0 opened the pull request:
This is a generalization of #122520 to other instructions with constrained result ranges.

Not sure whether this generalization is desirable, but I thought I would try it out.

jacobly0 committed:
[InstCombine][NFC] precommit tests for signed floor division

jacobly0 committed:
[InstCombine] fold unsigned predicates to signed on srem result

This allows optimization of more signed floor implementations when the
divisor is a known power of two to an arithmetic shift.

Proof for the implemented optimizations:
https://alive2.llvm.org/ce/z/j6C-Nz

Proof for the test cases:
https://alive2.llvm.org/ce/z/M_PBjw

jacobly0 committed:
[InstCombine][NFC] precommit tests for sign bit check canonicalization

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -3133,7 +3133,9 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
   if (ICmpInst::isUnsigned(Pred) && Add->hasNoSignedWrap() &&
       C.isNonNegative() && (C - *C2).isNonNegative() &&
-      computeConstantRange(X, /*ForSigned=*/true).add(*C2).isAllNonNegative())
+      computeConstantRange(X, Add, /*ForSigned=*/true)
```
I think `Add` is getting converted to a bool here and being used for `ForSigned` (then `/*ForSigned=*/true` is being used for `UseInstrInfo`).

jacobly0 replied:
Members take precedence over functions from an outer namespace, but I could use a different name for the member to be more clear.

goldsteinn commented on llvm/test/Transforms/InstCombine/smin-icmp.ll:
```diff
@@ -1004,18 +1004,17 @@ define void @eq_smin_v2i32_constant(<2 x i32> %y) {
-; CHECK-NEXT:    [[CMP5:%.*]] = icmp ult <2 x i32> [[COND]], splat (i32 10)
+; CHECK-NEXT:    [[CMP5:%.*]] = icmp sgt <2 x i32> [[Y:%.*]], splat (i32 -1)
 ; CHECK-NEXT:    call void @use_v2i1(<2 x i1> [[CMP5]])
-; CHECK-NEXT:    [[CMP6:%.*]] = icmp ult <2 x i32> [[COND]], splat (i32 11)
```
These two cases don't look profitable.

jacobly0 replied:
To be clear, this is `icmp ult (smin %Y, 5), 10` -> `icmp sgt %Y, -1`, is there a reason that removing an operation wouldn't be considered profitable here?

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7025,6 +7027,66 @@ static Instruction *canonicalizeICmpBool(ICmpInst &I,
+  if (*NegResult)
+    return new ICmpInst(ICmpInst::ICMP_SLT, X, ConstantInt::getNullValue(Ty));
+  return new ICmpInst(ICmpInst::ICMP_SGT, X, ConstantInt::getAllOnesValue(Ty));
+}
```
At least IMO, more useful than just building to just maximally constrain the `icmp`. I.e if you have `icmp ugt X, 10` and can actually narrow it to `icmp ugt X, 129`, thats better than `icmp slt X, 0`.

jacobly0 replied:
I do think it would be possible to update the downstream combines to handle a different canonical form.  My main worry is losing out in the backend on various shifting tricks and simpler sign checks that don't require loading a potentially large constant and which are localized to only needing to check one bit for large integers.  I'm sure this could be mitigated by moving the range checks and sign bit check preference into the backends that care, but it certainly does not appear to currently be the case.

Under the assumption that the issue is that more information could be gained by moving the comparison constant to one of the extremes, I'm still unclear on how to choose one side of the range as more beneficial.  It seems like some information is lost on each side, by reducing the known range on either a true or false branch, which is only gained back by recomputing ranges.  At that point, the choice of constant does not seem to affect the information you have, because the source range is always split into the same two ranges by any choice of constant.

If you want to limit it to only changing the constant of an existing icmp and not changing the signedness of the predicate, then that seems perfectly reasonable and potentially simple to implement.  If you have a specific definition of canonical other than preferring strict sign bit checks, I could also try to see if I could make it work.

goldsteinn replied:
I think is depends heavily on the use. I can see the case where if you are branching on the `cmp` then maximally constraining it in one direction leaves us with the least useful region on the other side. But in other use-cases whether only the `true` (or `false`) value matters we would almost certaintly want maximal constraint. I.e `(select (icmp ult X, 4), f(X), C)` is almost always better `(select (icmp sge X, 0), f(X), C)`.

I guess I can see sign-compare is a general middle-ground. Although I think this fold is liable to cause regressions in cases that it relaxes the constraints on a `cmp` that dominates some important expressions.

Truthfully not sure what the correct approach here is. I guess in some ideal case I would think some cost-function to decide which direction of the `cmp` is liable to benefit most from being maximally constrained, altough that might be overkill, especially in InstCombine.

jacobly0 replied:
Yep, that totally makes sense now, and I don't really have a better solution to propose at the moment.

Do you have a suggestion for how to proceed with my original `add (sdiv X, C), sext i1 (icmp ugt (srem X, C), smin)` => `ashr X, log2(C)` use case in the short term, and in a way that won't need to be reverted?  Is my original PR that only applies to `srem` seem restricted enough enough to avoid major regressions (especially given the relative rareness of `srem`), or should I revert back to something like my initial code which only applied where the full original `foldICmpSRemConstant` is applicable, just slightly improved based on what I have learned from these generalizations?

goldsteinn replied:
Are you talking about https://github.com/llvm/llvm-project/pull/122520? If so I think that can in without further changes.

jacobly0 committed:
[InstCombine] canonicalize sign bit checks

This is a generalization of #122520 to other instructions with
constrained result ranges.

jacobly0 commented:
Closing in favor of #122520 due to a concern about regressions caused by the canonicalization.
